clang/lib/Sema/SemaTemplateDeduction.cpp

```diff
@@ -3553,6 +3553,56 @@ static unsigned getPackIndexForParam(Sema &S,
   llvm_unreachable("parameter index would not be produced from template");
 }
 
+// if `Specialization` is a `CXXConstructorDecl` or `CXXConversionDecl`
+// we try to instantiate and update its explicit specifier after constraint
+// checking.
+static Sema::TemplateDeductionResult
+tryInstantiateExplicitSpecifier(Sema &S, FunctionDecl *Specialization,
+                                const MultiLevelTemplateArgumentList &SubstArgs,
+                                TemplateDeductionInfo &Info,
+                                FunctionTemplateDecl *FunctionTemplate,
+                                ArrayRef<TemplateArgument> DeducedArgs) {
+
+  const auto TryInstantiateExplicitSpecifierForSingleDecl =
+      [&](auto *ExplicitDecl) {
+        ExplicitSpecifier ExplicitSpecifier =
+            ExplicitDecl->getExplicitSpecifier();
+        Expr *const Expr = ExplicitSpecifier.getExpr();
+        if (!Expr) {
+          return Sema::TDK_Success;
+        }
+        if (!Expr->isValueDependent()) {
+          return Sema::TDK_Success;
+        }
+        // TemplateDeclInstantiator::InitFunctionInstantiation set the
+        // ActiveInstType to TemplateInstantiation, but we need
+        // to enable SFINAE when instantiating explicit specifier.
+        Sema::InstantiatingTemplate Inst(
+            S, Info.getLocation(), FunctionTemplate, DeducedArgs,
+            Sema::CodeSynthesisContext::DeducedTemplateArgumentSubstitution,
+            Info);
+        const auto Instantiated =
+            S.instantiateExplicitSpecifier(SubstArgs, ExplicitSpecifier);
+        if (Instantiated.isInvalid()) {
+          ExplicitDecl->setInvalidDecl(true);
+          return clang::Sema::TDK_SubstitutionFailure;
+        }
+        ExplicitDecl->setExplicitSpecifier(Instantiated);
+        return clang::Sema::TDK_Success;
+      };
+  Sema::TemplateDeductionResult DeductionResult = clang::Sema::TDK_Success;
+  if (CXXConstructorDecl *ConstructorDecl =
+          dyn_cast_or_null<CXXConstructorDecl>(Specialization)) {
```

Comment:
nit: use `dyn_cast` since we have ensured that `Specialization` is non-null at the caller site `FinishTemplateArgumentDeduction`.

Reply:
Thanks for reviewing ;)
Sadly, the type itself is required, because I need to call `CXXConversionDecl/CXXConstructorDecl::getExplicitSpecifier`/`setExplicitSpecifier` to get or set the explicit specifier.
I could not think of a better way to handle this without a template... 
If I put the function into `Sema` as a member function, I need to put two functions (one of them is the templated function) to `Sema`. Is that ok?

Reply:
Ah, I missed that. Hmm... Perhaps keep it inline and just extract the calls to the two functions (get and set ExplicitSpecifier) into their own little functions that 'do the right thing'.  So:

```
 auto getExplicitSpecifier(FunctionDecl *FD) {
   return isa<CXXConstructorDecl>(FD) ? cast<CXXConstructorDecl>(FD)->getExplicitSpecifier() : cast<CXXConversionDecl>(FD)->getExplicitSpecifier;
}
```

Reply:
Thanks for the advice, I’ve refactored the code. ;)